precompiles/assets-erc20/src/lib.rs

```diff
@@ -0,0 +1,444 @@
+		/*	let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+
+		// Read input.
+		input.expect_arguments(2)?;
+
+		let owner: H160 = input.read::<Address>()?.into();
+		let spender: H160 = input.read::<Address>()?.into();
+
+		// Fetch info.
+		let amount: U256 = {
+			let execution_address = Runtime::AddressMapping::into_account_id(context.address);
+			let asset_id: AssetIdOf<Runtime, Instance> =
+				Runtime::Precompiles::account_to_asset_id(execution_address)
+					.ok_or(error("non-assetId address"))?;
+
+			let owner: Runtime::AccountId = Runtime::AddressMapping::into_account_id(owner);
+			let spender: Runtime::AccountId = Runtime::AddressMapping::into_account_id(spender);
+
+			// Fetch info.
+			pallet_assets::Pallet::<Runtime, Instance>::allowance(asset_id, owner, spender).into()
+		};
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: EvmDataWriter::new().write(amount).build(),
+			logs: vec![],
+		})*/
+	}
+
+	fn approve(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_write_gas_cost())?;
+		gasometer.record_log_costs_manual(3, 32)?;
+
+		// Parse input.
+		input.expect_arguments(2)?;
+
+		let spender: H160 = input.read::<Address>()?.into();
+		let amount: U256 = input.read()?;
+
+		{
+			let execution_address = Runtime::AddressMapping::into_account_id(context.address);
+			let asset_id: AssetIdOf<Runtime, Instance> =
+				Runtime::Precompiles::account_to_asset_id(execution_address)
+					.ok_or(error("non-assetId address"))?;
+
+			let caller: Runtime::AccountId =
+				Runtime::AddressMapping::into_account_id(context.caller);
+			let spender: Runtime::AccountId = Runtime::AddressMapping::into_account_id(spender);
+			let amount = Self::u256_to_amount(amount)?;
+
+			// Dispatch call (if enough gas).
+			// We first cancel any existing approvals
+			// Since we cannot check storage, we need to execute this call without knowing whether
+			// another approval exists already.
+			// But we know that if no approval exists we should get "Unknown"
+			// Allowance() should be checked instead of doing this Result matching
+			let used_gas = match RuntimeHelper::<Runtime>::try_dispatch(
+				<<Runtime as frame_system::Config>::Call as Dispatchable>::Origin::root(),
+				pallet_assets::Call::<Runtime, Instance>::force_cancel_approval(
+					asset_id,
+					Runtime::Lookup::unlookup(caller),
+					Runtime::Lookup::unlookup(spender.clone()),
+				),
+				gasometer.remaining_gas()?,
+			) {
+				Ok(gas_used) => Ok(gas_used),
+				Err(ExitError::Other(e)) => {
+					// This would mean there is not an existing approval
+					// We convert this case to 0 gas used
+					// We could also convert it to one DB read
```

Comment:
Yes, we need to still charge for the checks we did to discover that there was no such asset.